server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java

```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.filter;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.BulkScorer;
+import org.apache.lucene.search.ConstantScoreQuery;
+import org.apache.lucene.search.IndexOrDocValuesQuery;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.IndexSortSortedNumericDocValuesRangeQuery;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.PointRangeQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.common.lucene.Lucene;
+
+import java.io.IOException;
+import java.util.function.BiConsumer;
+import java.util.function.IntPredicate;
+
+/**
+ * Adapts a Lucene {@link Query} to the behaviors used be the
+ * {@link FiltersAggregator}. In general we try to delegate to {@linkplain Query}
+ * when we don't have 
+ */
+public abstract class QueryToFilterAdapter {
```

Comment:
Two things bothered me about it:
1. The base class would need a `<>` operator which is kind of a pain because it's really just an implementation detail.
2. MatchNone doesn't really want any of the "default" stuff. It'd fully override it. MatchAll is pretty similar. But those are the only two like that.

I'd certainly be willing to try and do all the merging and see what it looks like.